<a href="https://colab.research.google.com/github/Diego-Rodriguez16/AlgorithmsUN2024II/blob/main/Lab%203/dirodriguezd_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

4 10
14

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

11

In [ ]:
!./APlusB < input2.txt

20

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

14 55
69


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

3
4 5 15
75


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
3
10 300 700000

Writing input4.txt


In [ ]:
%%writefile output4.txt
210000000

Writing output4.txt


In [ ]:
%%writefile input5.txt
8
40 888 47 999 52 4000 7878 8989

Writing input5.txt


In [ ]:
%%writefile output5.txt
70815342

Writing output5.txt


In [ ]:
# Stress Test case 1
%%writefile input6.txt
6
1000000000 234000000 200000000 9000 200000000 892000000

Writing input6.txt


In [ ]:
%%writefile output6.txt
892000000000000000

Writing output6.txt


In [ ]:
#Stress Test case 2
%%writefile input7.txt
5
-1000000000 -234000000 -200000000 -9000 -200000000

Writing input7.txt


In [ ]:
%%writefile output7.txt
234000000000000000

Writing output7.txt


In [ ]:
#Stress Test case 3
%%writefile input8.txt
5
10000000 4 200000 9000 1

Writing input8.txt


In [ ]:
%%writefile output8.txt
2000000000000

Writing output8.txt


In [ ]:
%%writefile input9.txt
1
8

Writing input9.txt


In [ ]:
%%writefile output9.txt
0

Writing output9.txt


In [ ]:
%%writefile input10.txt
9
4 78 98 52 444 6333 2111 010101 48

Writing input10.txt


In [ ]:
%%writefile output10.txt
63969633

Writing output10.txt


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {4..10}
do
    echo "Ejecutando test $i..."
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt

    # Compara los resultados
    if grep -xvFf output$i.txt ctest$i.txt > /dev/null; then
        echo "-- Test $i: Falló --"
        echo "-- Salida obtenida: --"
        cat ctest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i: Pasó"
    fi
done


Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 4...
Test 4: Pasó
Ejecutando test 5...
Test 5: Pasó
Ejecutando test 6...
Test 6: Pasó
Ejecutando test 7...
Test 7: Pasó
Ejecutando test 8...
Test 8: Pasó
Ejecutando test 9...
Test 9: Pasó
Ejecutando test 10...
Test 10: Pasó


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m2.136s
user	4m58.921s
sys	0m0.089s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m1.565s
user	4m58.903s
sys	0m0.095s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177091


In [ ]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177091


1.474976569326082e-09

In [ ]:
#Using the real reported: 5m2 --> 302
302/200003177091

1.5099760133439889e-09

In [ ]:
# Estimated number of instructions by second

200003177091/295

677976871.4949152

In [ ]:
#Using the real reported: 5m2 --> 302
200003177091/302

662262175.7980132

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.301s
user	0m1.265s
sys	0m0.002s


## Python

In [ ]:
%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(nums):
    product_max = 0  # Producto máximo inicializado
    product_temp = 0  # Producto temporal para comparar
    length = len(nums)  # Longitud de la lista de números

    # Dos bucles anidados para revisar todos los pares posibles
    for i in range(length):
        for j in range(length):
            if i != j:  # Evita comparar el mismo número consigo mismo
                product_temp = nums[i] * nums[j]  # Producto de los dos números
                if product_temp > product_max:  # Si el producto temporal es mayor, actualiza el máximo
                    product_max = product_temp

    return product_max  # Devuelve el producto máximo encontrado

# Entrada de datos
count = int(input())  # Número de elementos en la lista
nums = [int(x) for x in input().split()]  # Lista de números

# Llama a la función y muestra el resultado
print(MaxPairwiseProduct(nums))



Writing max_pairwise_product_long_full.py


In [ ]:
# Verify
!time python3 max_pairwise_product_long_full.py < input13.txt

120989000

real	0m20.661s
user	0m20.417s
sys	0m0.018s


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product < input13.txt

120989000

real	0m0.548s
user	0m0.544s
sys	0m0.001s


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.409s
user	0m0.401s
sys	0m0.004s


##Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

    long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = first + 1; second < n; ++second) {                              //2n + (n-1)n
            execints += 2;

            aux_product = (long long)numbers[first] * numbers[second];                    //(n-1)n/2
            execints++;

            execints++;
            if (max_product < aux_product) {                                              //(n-1)n/2
                max_product = aux_product;                                               //[0, (n-1)n/2] upper bound
                execints++;
            }
        }
        execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 4n + [2(n-1)n + 1, 2.5(n-1)n) upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints = 0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];                              //n
        execints++;
    }

    execints++;
    int M = MaxPairwiseProduct(numbers, execints);

    execints++;
    std::cout << execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 7n + [2(n-1)n + 1, 2.5(n-1)n) upper bound


Writing max_pairwise_product_long_execints.cpp


In [ ]:
# Estimated number of instructions
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input13.txt

242181051


In [ ]:
# Estimated number of instructions by second
242181051/0.409

592129709.0464548

# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [ ]:
%%writefile max_pairwise_product_sort_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

long long maxPairwiseProduct(const vector<int>& arr) {
    vector<int> sortedArr = arr;
    sort(sortedArr.begin(), sortedArr.end());
    long long max1 = static_cast<long long>(sortedArr[sortedArr.size() - 1]);
    long long max2 = static_cast<long long>(sortedArr[sortedArr.size() - 2]);

    return max1 * max2;
}

int main() {
    int n;
    cin >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }

    long long result = maxPairwiseProduct(arr);
    cout << result;
    return 0;
}


Overwriting max_pairwise_product_sort_cpp.cpp


In [ ]:
%%writefile max_pairwise_product_sort.py
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

n = int(input())
numbers = [int(i) for i in input().split()]
print(max_pairwise_product_sort(numbers))

Overwriting max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_sort_cpp.cpp -o max_pairwise_product_sort_cpp
for i in {1..2}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_sort < input$i.txt > ctest$i.txt

    if grep -xvFf output$i.txt ctest$i.txt > /dev/null; then
        echo "-- Test $i (C++): Falló --"
        echo "-- Salida obtenida: --"
        cat ctest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (C++): Pasó"
    fi
done

for i in {1..2}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_sort.py < input$i.txt > ptest$i.txt

    # Comparar resultados
    if grep -xvFf output$i.txt ptest$i.txt > /dev/null; then
        echo "-- Test $i (Python): Falló --"
        echo "-- Salida obtenida: --"
        cat ptest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (Python): Pasó"
    fi
done


Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
Test 1 (C++): Pasó
Ejecutando test 2 con C++...
Test 2 (C++): Pasó
Ejecutando test 1 con Python...
Test 1 (Python): Pasó
Ejecutando test 2 con Python...
Test 2 (Python): Pasó


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_sort_cpp < input3.txt > cplusplusOutput3.txt
!grep -xvFf cplusplusOutput3.txt output3.txt

!time python3 max_pairwise_product_sort.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt



real	0m0.005s
user	0m0.001s
sys	0m0.003s

real	0m0.037s
user	0m0.023s
sys	0m0.014s


20 stress tests

In [ ]:
import subprocess
import random

!g++ -o max_pairwise_product_sort_cpp max_pairwise_product_sort_cpp.cpp

def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    return secondLargest * largest

def generate_test_case(max_size=10000, max_value=10000000):
    size = random.randint(2, max_size)
    numbers = [random.randint(0, max_value) for _ in range(size)]
    return size, numbers

for test_id in range(20):
    size, numbers = generate_test_case()
    input_data = f"{size}\n{' '.join(map(str, numbers))}\n"

    # Mostrar el input data
    print(f"Test {test_id + 1}:")
    print(f"Input data: {input_data.strip()}")

    # Ejecutar el código C++ y capturar salida
    cpp_result = subprocess.run(
        ["./max_pairwise_product_sort_cpp"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    # Ejecutar el código Python y capturar salida
    py_result = subprocess.run(
        ["python3", "max_pairwise_product_sort.py"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    efficient = efficientSolution(numbers)

    # Validar que los resultados sean iguales
    if int(cpp_result) != efficient or int(py_result) != efficient:
        print(f"❌ Test {test_id + 1} failed!")
        print(f"C++ Output: {cpp_result}")
        print(f"Python Output: {py_result}")
        break
    else:
        print(f"✅ Test {test_id + 1} passed!")
        print(f"Output c++: {cpp_result}")
        print(f"Output python: {py_result}")
        print(f"Output efficient: {efficient}\n")

Test 1:
Input data: 3380
2099902 3225331 8463227 1434665 5664665 8506111 2553604 8283970 5376555 7332448 5609848 2282112 7307342 86378 5846644 6250786 5939733 918663 2410017 2043014 9494648 1214717 5858672 1399555 2639322 418927 7640972 5169765 8874643 364622 703257 9593995 8028896 6829225 2276848 6979690 274496 7720881 2083996 1774673 3218403 574654 9184315 8449146 2899089 4971392 6357022 4044143 9249411 4578673 8758723 4835832 2592100 8037256 5791002 2433047 9566021 3127 3699960 6423562 8485030 3285867 5333352 7926334 1555386 6679608 3460371 9426287 4844914 4148416 8149543 8193385 3611288 762381 3058807 5964230 6134237 3576934 3126797 5415580 2463472 8813404 9420134 7567468 6096913 9738125 6026042 809237 9540463 3203337 2635577 58122 3967676 8311666 4620551 6331567 6861480 7929644 1622163 1122539 6274791 7116220 9239431 2573337 8020291 2644662 1728215 4468572 5673372 2516681 2203625 6478182 3950389 8537690 4755402 5017907 4588699 3417135 2023265 9885965 6354807 2917904 9584767 217217

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_cpp_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

long long maxPairwiseProduct(vector<int>& arr, long long &execints) {
    execints++;
    sort(arr.begin(), arr.end());
    execints += static_cast<long long>(arr.size() * log2(arr.size()));
    long long max1 = static_cast<long long>(arr[arr.size() - 1]);
    execints++;
    long long max2 = static_cast<long long>(arr[arr.size() - 2]);
    execints++;
    execints++;
    return max1 * max2;
}

int main() {
    long long execints = 0LL;
    int n;
    execints++;
    cin >> n;
    execints++;
    vector<int> arr(n);
    execints += n;
    execints++;
    for (int i = 0; i < n; ++i) {
        execints += 2;
        cin >> arr[i];
        execints++;
    }
    execints++;
    long long result = maxPairwiseProduct(arr, execints);
    execints++;
    cout << "Cantidad de instrucciones: " << execints << "\n";

    return 0;
}

Overwriting max_pairwise_product_sort_cpp_execints.cpp


In [ ]:
%%writefile max_pairwise_product_sort_execints.py
def max_pairwise_product_sort(numbers, execints):
    execints[0] += 1  # Sort call
    numbers.sort(reverse=True)  # O(n log n)
    execints[0] += len(numbers) * (len(numbers).bit_length())  # Approximation of sort complexity

    execints[0] += 1  # Accessing numbers[0]
    max1 = numbers[0]

    execints[0] += 1  # Accessing numbers[1]
    max2 = numbers[1]

    execints[0] += 1  # Multiplication
    return max1 * max2

execints = [0]  # Use a list to pass by reference

execints[0] += 1  # Reading n
n = int(input())

execints[0] += 1  # Reading and parsing numbers
numbers = [int(i) for i in input().split()]
execints[0] += n  # List comprehension loop

execints[0] += 1  # Function call
result = max_pairwise_product_sort(numbers, execints)

execints[0] += 1  # Printing result
print("Cantidad de instrucciones: ",execints[0])


Overwriting max_pairwise_product_sort_execints.py


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_sort_cpp_execints.cpp -o max_pairwise_product_sort_cpp_execints
for i in {1..3}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_sort_cpp_execints < input$i.txt
done

for i in {1..3}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_sort_execints.py < input$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
Cantidad de instrucciones: 82
Ejecutando test 2 con C++...
Cantidad de instrucciones: 82
Ejecutando test 3 con C++...
Cantidad de instrucciones: 48
Ejecutando test 1 con Python...
Cantidad de instrucciones:  58
Ejecutando test 2 con Python...
Cantidad de instrucciones:  58
Ejecutando test 3 con Python...
Cantidad de instrucciones:  32


Calculate estimated time by instruction and number of instructions by second

In [ ]:
print("----- C++:------")
!g++ max_pairwise_product_sort_cpp_execints.cpp -o max_pairwise_product_sort_cpp_execints
!time ./max_pairwise_product_sort_cpp_execints< input11.txt
print("\n-----Python:------")
!time python3 max_pairwise_product_sort_execints.py < input11.txt


----- C++:------
Cantidad de instrucciones: 4321937

real	0m0.118s
user	0m0.114s
sys	0m0.003s

-----Python:------
Cantidad de instrucciones:  3800008

real	0m0.104s
user	0m0.073s
sys	0m0.026s


In [ ]:
# Estimated time by instruction
print("Estimated time by instruction C++:",0.118/4321937)
print("Estimated time by instruction Python:",0.104/3800008)
# Estimated number of instructions by second
print("Estimated number of instructions by second C++:",4321937/0.118)
print("Estimated number of instructions by second Python:",3800008/0.104)

Estimated time by instruction C++: 2.730257289729119e-08
Estimated time by instruction Python: 2.7368363435024346e-08
Estimated number of instructions by second C++: 36626584.74576271
Estimated number of instructions by second Python: 36538538.461538464


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_efficient.cpp
#include <iostream>
#include <vector>
using namespace std;

long long maxPairwiseProduct(const vector<int>& arr) {
    int n = arr.size(); // 1 instrucción
    int max1 = -1, max2 = -1; // 2 instrucciones

    // Encontrar el mayor elemento
    for (int i = 0; i < n; ++i) { // n instrucciones para el bucle
        if (max1 == -1 || arr[i] > arr[max1]) { // 2n comparaciones
            max2 = max1; // n asignaciones
            max1 = i; // n asignaciones
        } else if (max2 == -1 || arr[i] > arr[max2]) { // 2n comparaciones
            max2 = i; // n asignaciones
        }
    }

    return static_cast<long long>(arr[max1]) * arr[max2]; // 1 multiplicación y retorno
}

int main() {
    int n;
    cin >> n; // 1 lectura

    vector<int> arr(n); // 1 inicialización
    for (int i = 0; i < n; ++i) { // n asignaciones
        cin >> arr[i]; // n lecturas
    }

    long long result = maxPairwiseProduct(arr); // 1 llamada de función
    cout << result; // 1 escritura
    return 0;
}


Writing max_pairwise_product_efficient.cpp


In [ ]:
%%writefile max_pairwise_product_efficient.py
def max_pairwise_product(numbers):
    max1, max2 = -1, -1  # 2 asignaciones

    # Encontrar el mayor elemento
    for i in range(len(numbers)):  # n iteraciones
        if max1 == -1 or numbers[i] > numbers[max1]:  # 2n comparaciones
            max2 = max1  # n asignaciones
            max1 = i  # n asignaciones
        elif max2 == -1 or numbers[i] > numbers[max2]:  # 2n comparaciones
            max2 = i  # n asignaciones

    return numbers[max1] * numbers[max2]  # 1 multiplicación

n = int(input())  # 1 lectura
numbers = list(map(int, input().split()))  # n asignaciones y conversiones
print(max_pairwise_product(numbers))  # 1 llamada de función y escritura


Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2



In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_efficient.cpp -o max_pairwise_product_efficient
for i in {1..2}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_efficient < input$i.txt > ctest$i.txt

    if grep -xvFf output$i.txt ctest$i.txt > /dev/null; then
        echo "-- Test $i (C++): Falló --"
        echo "-- Salida obtenida: --"
        cat ctest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (C++): Pasó"
    fi
done

for i in {1..2}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_efficient.py < input$i.txt > ptest$i.txt

    # Comparar resultados
    if grep -xvFf output$i.txt ptest$i.txt > /dev/null; then
        echo "-- Test $i (Python): Falló --"
        echo "-- Salida obtenida: --"
        cat ptest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (Python): Pasó"
    fi
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
Test 1 (C++): Pasó
Ejecutando test 2 con C++...
Test 2 (C++): Pasó
Ejecutando test 1 con Python...
Test 1 (Python): Pasó
Ejecutando test 2 con Python...
Test 2 (Python): Pasó


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_efficient < input3.txt > cplusplusOutput3.txt
!grep -xvFf cplusplusOutput3.txt output3.txt

!time python3 max_pairwise_product_efficient.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt


real	0m0.003s
user	0m0.001s
sys	0m0.002s

real	0m0.035s
user	0m0.023s
sys	0m0.011s


20 stress tests

In [ ]:
import subprocess
import random

!g++ -o max_pairwise_product_efficient max_pairwise_product_efficient.cpp

def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    return secondLargest * largest

def generate_test_case(max_size=10000, max_value=10000000):
    size = random.randint(2, max_size)
    numbers = [random.randint(0, max_value) for _ in range(size)]
    return size, numbers

for test_id in range(20):
    size, numbers = generate_test_case()
    input_data = f"{size}\n{' '.join(map(str, numbers))}\n"

    # Mostrar el input data
    print(f"Test {test_id + 1}:")
    print(f"Input data: {input_data.strip()}")

    # Ejecutar el código C++ y capturar salida
    cpp_result = subprocess.run(
        ["./max_pairwise_product_efficient"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    # Ejecutar el código Python y capturar salida
    py_result = subprocess.run(
        ["python3", "max_pairwise_product_efficient.py"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    efficient = efficientSolution(numbers)

    # Validar que los resultados sean iguales
    if int(cpp_result) != efficient or int(py_result) != efficient:
        print(f"❌ Test {test_id + 1} failed!")
        print(f"C++ Output: {cpp_result}")
        print(f"Python Output: {py_result}")
        break
    else:
        print(f"✅ Test {test_id + 1} passed!")
        print(f"Output c++: {cpp_result}")
        print(f"Output python: {py_result}")
        print(f"Output efficient: {efficient}\n")

Test 1:
Input data: 898
249922 8566418 5750769 9480723 7512508 3616652 8370459 9800937 6131431 1285587 5911876 3059596 2627500 1706437 9278683 1265007 4387041 4279288 3684736 9957296 4174214 5091739 9713009 3675327 7749384 4740460 8475444 7368995 2497766 582221 7790894 7074262 9029860 8009174 5366407 9533471 1152808 1064204 453454 8107667 3591380 1045317 6675581 1866541 2920347 7604517 9894740 5033707 624198 3068449 3501885 6949370 3282019 2818627 3081074 6426288 3795604 3987309 9557237 2918789 5821697 2242367 839189 5747069 2593270 4784132 3027889 9023267 2353708 7135405 2899697 1055791 3145175 813305 7537829 2227395 533653 309470 4966797 5502470 264142 3710245 5673953 9645860 336677 6747049 3791405 7636978 3136613 2279296 5542099 756514 1964168 5514243 3518722 9311393 8304053 5351204 2360069 7371302 7015538 1748301 4833764 8010149 2979065 3394135 8178905 4331834 355663 9676542 1356697 223163 7622094 2749658 5323555 7777262 1698061 8074354 5504144 5753944 5404171 7547935 6200633 99791

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_efficient_execints.cpp
#include <iostream>
#include <vector>
using namespace std;

long long maxPairwiseProduct(const vector<int>& arr, long long &execints) {
    int n = arr.size();  // 1 instrucción
    execints++;

    int max1 = -1, max2 = -1;  // 2 instrucciones
    execints += 2;

    // Encontrar el mayor elemento
    execints++;
    for (int i = 0; i < n; ++i) {  // n instrucciones para el bucle
        execints++;

        execints++;
        if (max1 == -1 || arr[i] > arr[max1]) {  // 2n comparaciones
            max2 = max1;  // n asignaciones
            execints++;

            max1 = i;  // n asignaciones
            execints++;
        } else {
            execints++;
            if (max2 == -1 || arr[i] > arr[max2]) {  // 2n comparaciones
                max2 = i;  // n asignaciones
                execints++;
            }
        }
    }

    execints++;
    return static_cast<long long>(arr[max1]) * arr[max2];  // 1 multiplicación y retorno
}

int main() {
    long long execints = 0LL;

    int n;
    cin >> n;  // 1 lectura
    execints++;

    vector<int> arr(n);  // 1 inicialización
    execints++;

    execints++;
    for (int i = 0; i < n; ++i) {  // n asignaciones
        execints++;

        cin >> arr[i];  // n lecturas
        execints++;
    }

    execints++;
    long long result = maxPairwiseProduct(arr, execints);  // 1 llamada de función

    execints++;
    cout << execints << "\n";  // 1 escritura
    return 0;
}


Writing max_pairwise_product_efficient_execints.cpp


In [ ]:
%%writefile max_pairwise_product_efficient_execints.py
def max_pairwise_product(numbers, execints):
    max1, max2 = -1, -1  # 2 asignaciones
    execints += 2

    # Encontrar el mayor elemento
    execints += 1
    for i in range(len(numbers)):  # n iteraciones
        execints += 1

        execints += 1
        if max1 == -1 or numbers[i] > numbers[max1]:  # 2n comparaciones
            max2 = max1  # n asignaciones
            execints += 1

            max1 = i  # n asignaciones
            execints += 1
        elif max2 == -1 or numbers[i] > numbers[max2]:  # 2n comparaciones
            max2 = i  # n asignaciones
            execints += 1

    execints += 1
    return numbers[max1] * numbers[max2]  # 1 multiplicación

n = int(input())  # 1 lectura
execints = 1  # cuenta la lectura de n
numbers = list(map(int, input().split()))  # n asignaciones y conversiones
execints += len(numbers)

result = max_pairwise_product(numbers, execints)  # 1 llamada de función
execints += 1

print(execints)  # Imprimir el número total de instrucciones


Writing max_pairwise_product_efficient_execints.py


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_efficient_execints.cpp -o max_pairwise_product_efficient_execints
for i in {1..3}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_efficient_execints < input$i.txt
done

for i in {1..3}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_efficient_execints.py < input$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
70
Ejecutando test 2 con C++...
62
Ejecutando test 3 con C++...
45
Ejecutando test 1 con Python...
12
Ejecutando test 2 con Python...
12
Ejecutando test 3 con Python...
8


Calculate estimated time by instruction and number of instructions by second

In [ ]:
print("----- C++:------")
!g++ max_pairwise_product_efficient_execints.cpp -o max_pairwise_product_efficient_execints
!time ./max_pairwise_product_efficient_execints< input11.txt
print("\n-----Python:------")
!time python3 max_pairwise_product_efficient_execints.py < input11.txt

----- C++:------
1200010

real	0m0.077s
user	0m0.075s
sys	0m0.001s

-----Python:------
200002

real	0m0.190s
user	0m0.128s
sys	0m0.038s


In [ ]:
# Estimated time by instruction
print("Estimated time by instruction C++:",0.077/1200010)
print("Estimated time by instruction Python:",0.190/200002)
# Estimated number of instructions by second
print("Estimated number of instructions by second C++:",1200010/0.077)
print("Estimated number of instructions by second Python:",200002/0.190)

Estimated time by instruction C++: 6.416613194890043e-08
Estimated time by instruction Python: 9.49990500094999e-07
Estimated number of instructions by second C++: 15584545.454545455
Estimated number of instructions by second Python: 1052642.105263158


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [ ]:
%%writefile max_pairwise_product_super_efficient.cpp
#include <iostream>
using namespace std;

int main() {
    int n;
    cin >> n; // 1 lectura

    long long max1 = -1, max2 = -1; // 2 inicializaciones
    for (int i = 0; i < n; ++i) { // n iteraciones
        int num;
        cin >> num; // n lecturas
        if (num > max1) { // 2n comparaciones
            max2 = max1; // n asignaciones
            max1 = num; // n asignaciones
        } else if (num > max2) { // 2n comparaciones
            max2 = num; // n asignaciones
        }
    }

    cout << max1 * max2; // 1 multiplicación y 1 escritura
    return 0; // 1 retorno
}


Writing max_pairwise_product_super_efficient.cpp


Test with inputs 1 and 2




In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_super_efficient.cpp -o max_pairwise_product_super_efficient
for i in {1..2}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_super_efficient < input$i.txt > ctest$i.txt

    if grep -xvFf output$i.txt ctest$i.txt > /dev/null; then
        echo "-- Test $i (C++): Falló --"
        echo "-- Salida obtenida: --"
        cat ctest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (C++): Pasó"
    fi
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
Test 1 (C++): Pasó
Ejecutando test 2 con C++...
Test 2 (C++): Pasó


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_super_efficient < input3.txt > cplusplusOutput3.txt
!grep -xvFf cplusplusOutput3.txt output3.txt


real	0m0.006s
user	0m0.001s
sys	0m0.002s


20 stress tests

In [ ]:
import subprocess
import random

!g++ -o max_pairwise_product_super_efficient max_pairwise_product_super_efficient.cpp

def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    return secondLargest * largest

def generate_test_case(max_size=10000, max_value=10000000):
    size = random.randint(2, max_size)
    numbers = [random.randint(0, max_value) for _ in range(size)]
    return size, numbers

for test_id in range(20):
    size, numbers = generate_test_case()
    input_data = f"{size}\n{' '.join(map(str, numbers))}\n"

    # Mostrar el input data
    print(f"Test {test_id + 1}:")
    print(f"Input data: {input_data.strip()}")

    # Ejecutar el código C++ y capturar salida
    cpp_result = subprocess.run(
        ["./max_pairwise_product_super_efficient"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    efficient = efficientSolution(numbers)

    # Validar que los resultados sean iguales
    if int(cpp_result) != efficient:
        print(f"❌ Test {test_id + 1} failed!")
        print(f"C++ Output: {cpp_result}")
        print(f"Output efficient: {efficient}\n")
        break
    else:
        print(f"✅ Test {test_id + 1} passed!")
        print(f"Output c++: {cpp_result}\n")

Test 1:
Input data: 963
7880323 379078 3686407 701540 4290924 5602627 9067968 585951 6816312 7817981 4326337 5795079 6443684 3471042 877873 9837256 1785344 547894 1413752 4690513 4155341 8394950 546081 8826921 1335371 8611004 5043177 8645825 963871 2436803 9370888 9262887 9508595 3196960 2700668 5274070 1698573 798928 1565315 8754540 3628441 1842310 8093139 4001401 5209537 6996227 902307 6240609 6459238 9378165 1252830 5527382 8560651 5535979 4807044 8943444 5180912 8759209 5426931 7342639 3001867 1039317 3138528 304208 3317343 7596541 9412321 6194525 6016617 7069815 4921552 154335 999508 5319712 5848032 4393001 7563401 7090842 1603474 9925260 5441503 386400 5754747 3583618 7886227 1604195 3247489 6173240 9698401 1953755 4133305 3939259 824655 1008974 58595 7717734 2146509 9787308 12169 8273810 9715134 5850284 2684302 5845381 7667278 5768284 6428168 1293854 6190430 8830431 1639983 2861308 2297638 660104 3847017 2648761 1764067 914749 8415905 6817312 9537212 573698 9834046 7498489 31230

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_super_efficient_execints.cpp
#include <iostream>
using namespace std;

int main() {
    long long execints = 0LL;

    int n;
    cin >> n;  // 1 lectura
    execints++;

    long long max1 = -1, max2 = -1;  // 2 inicializaciones
    execints += 2;

    for (int i = 0; i < n; ++i) {  // n iteraciones
        execints++;

        int num;
        cin >> num;  // n lecturas
        execints++;

        if (num > max1) {  // 2n comparaciones
            max2 = max1;  // n asignaciones
            execints++;

            max1 = num;  // n asignaciones
            execints++;
        } else if (num > max2) {  // 2n comparaciones
            max2 = num;  // n asignaciones
            execints++;
        }
    }

    long long result = max1 * max2;  // 1 multiplicación y 1 escritura
    execints++;

    cout << execints << "\n";  // 1 escritura

    return 0;  // 1 retorno
}


Overwriting max_pairwise_product_super_efficient_execints.cpp


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_super_efficient_execints.cpp -o max_pairwise_product_super_efficient_execints
for i in {1..3}
do
    echo "Ejecutando test $i con C++..."
    ./max_pairwise_product_super_efficient_execints < input$i.txt
done


Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con C++...
44
Ejecutando test 2 con C++...
27
Ejecutando test 3 con C++...
24


Calculate estimated time by instruction and number of instructions by second

In [ ]:
print("----- C++:------")
!g++ max_pairwise_product_super_efficient_execints.cpp -o max_pairwise_product_super_efficient_execints
!time ./max_pairwise_product_super_efficient_execints< input11.txt

----- C++:------
800004

real	0m0.092s
user	0m0.082s
sys	0m0.006s


In [ ]:
# Estimated time by instruction
print("Estimated time by instruction C++:",0.092/800004)
# Estimated number of instructions by second
print("Estimated number of instructions by second C++:",800004/0.092)

Estimated time by instruction C++: 1.1499942500287498e-07
Estimated number of instructions by second C++: 8695695.652173914


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_super_efficient.py
n = int(input())  # 1 lectura
nums = list(map(int, input().split()))  # Convierte la línea en una lista de enteros

max1, max2 = -1, -1  # 2 inicializaciones
for num in nums:  # n iteraciones
    if num > max1:  # 2n comparaciones
        max2 = max1  # n asignaciones
        max1 = num  # n asignaciones
    elif num > max2:  # 2n comparaciones
        max2 = num  # n asignaciones

print(max1 * max2)  # 1 multiplicación y 1 escritura


Overwriting max_pairwise_product_super_efficient.py


Test with inputs 1 and 2




In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
for i in {1..2}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_super_efficient.py < input$i.txt > ptest$i.txt

    # Comparar resultados
    if grep -xvFf output$i.txt ptest$i.txt > /dev/null; then
        echo "-- Test $i (Python): Falló --"
        echo "-- Salida obtenida: --"
        cat ptest$i.txt
        echo "-- Salida esperada: --"
        cat output$i.txt
    else
        echo "Test $i (Python): Pasó"
    fi
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con Python...
Test 1 (Python): Pasó
Ejecutando test 2 con Python...
Test 2 (Python): Pasó


Test and time it with input 3

In [ ]:
!time python3 max_pairwise_product_super_efficient.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt


real	0m0.036s
user	0m0.028s
sys	0m0.008s


20 stress tests

In [ ]:
import subprocess
import random

def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    return secondLargest * largest

def generate_test_case(max_size=10000, max_value=10000000):
    size = random.randint(2, max_size)
    numbers = [random.randint(0, max_value) for _ in range(size)]
    return size, numbers

for test_id in range(20):
    size, numbers = generate_test_case()
    input_data = f"{size}\n{' '.join(map(str, numbers))}\n"

    # Mostrar el input data
    print(f"Test {test_id + 1}:")
    print(f"Input data: {input_data.strip()}")

    # Ejecutar el código Python y capturar salida
    py_result = subprocess.run(
        ["python3", "max_pairwise_product_efficient.py"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    efficient = efficientSolution(numbers)

    # Validar que los resultados sean iguales
    if int(py_result) != efficient:
        print(f"❌ Test {test_id + 1} failed!")
        print(f"Python Output: {py_result}")
        print(f"Output efficient: {efficient}\n")
        break
    else:
        print(f"✅ Test {test_id + 1} passed!")
        print(f"Output python: {py_result}\n")

Test 1:
Input data: 7464
6522459 5970051 6272417 334817 4022799 6112878 8178277 4084905 6975546 972676 568235 2660359 8472038 4500169 7091021 692113 644886 2524857 8168270 1292235 5428541 2710828 8135450 2749888 116349 9111901 2783731 640045 9506596 5234670 6170807 9328830 6477794 2066119 8332897 7249816 2736906 4780903 2585168 8021778 6971960 275224 2140940 2437739 7535565 5316667 6243661 1755105 6136367 1387387 8685229 3596841 6830652 4592986 1450536 9699079 7774850 5542282 3624475 1749442 7902276 8695348 5181542 4771712 2786231 8815825 6370639 7471310 9331743 9877334 451567 3500007 4680884 2669196 1321900 8449322 8092878 1134878 5338144 1837280 1748969 1536707 7270519 1540407 6693974 29708 6173864 5015147 4885258 7021483 1416480 6134247 4436430 2944424 8813792 5242752 4193771 1705558 5755138 8368313 9752304 2214587 3406632 6252713 1105503 2462700 3312409 5131387 6263843 7062727 7049474 995964 7354979 2829674 1148915 4740850 8888954 451818 1191450 4051783 7782183 1294930 251762 43108

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_super_efficient_execints.py
n = int(input())  # 1 lectura
execints = 1  # 1 instrucción para leer el valor de n

nums = list(map(int, input().split()))  # Convierte la línea en una lista de enteros
execints += 1  # 1 para la lectura de la línea y n para mapear los números

max1, max2 = -1, -1  # 2 inicializaciones
execints += 2  # 2 inicializaciones

for num in nums:  # n iteraciones
    execints += 1  # 1 instrucción por iteración del bucle
    if num > max1:  # 2n comparaciones
        execints += 2  # 2 comparaciones
        max2 = max1  # n asignaciones
        max1 = num  # n asignaciones
        execints += 2  # 2 asignaciones
    elif num > max2:  # 2n comparaciones
        execints += 2  # 2 comparaciones
        max2 = num  # n asignaciones
        execints += 1  # 1 asignación

Result = (max1 * max2)  # 1 multiplicación y 1 escritura
execints += 2  # 1 para la multiplicación y 1 para la escritura

print(execints)  # Mostrar la cantidad total de instrucciones ejecutadas


Overwriting max_pairwise_product_super_efficient_execints.py


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
for i in {1..3}
do
    echo "Ejecutando test $i con Python..."
    python3 max_pairwise_product_super_efficient_execints.py < input$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Ejecutando test 1 con Python...
56
Ejecutando test 2 con Python...
23
Ejecutando test 3 con Python...
30


Calculate estimated time by instruction and number of instructions by second

In [265]:
print("\n-----Python:------")
!time python3 max_pairwise_product_super_efficient_execints.py < input11.txt


-----Python:------
1000006

real	0m0.168s
user	0m0.147s
sys	0m0.019s


In [266]:
# Estimated time by instruction
print("Estimated time by instruction Python:",0.168/1000006)
# Estimated number of instructions by second
print("Estimated number of instructions by second Python:",1000006/0.168)

Estimated time by instruction Python: 1.6799899200604797e-07
Estimated number of instructions by second Python: 5952416.666666666


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [269]:
import subprocess
import random

!g++ -o max_pairwise_product_sort_cpp max_pairwise_product_sort_cpp.cpp
!g++ -o max_pairwise_product_super_efficient max_pairwise_product_super_efficient.cpp

def generate_test_case(max_size=10000, max_value=10000000):
    size = random.randint(2, max_size)
    numbers = [random.randint(0, max_value) for _ in range(size)]
    return size, numbers

for test_id in range(100):
    size, numbers = generate_test_case()
    input_data = f"{size}\n{' '.join(map(str, numbers))}\n"

    # Ejecutar el código C++ y capturar salida
    cpp_result = subprocess.run(
        ["./max_pairwise_product_sort_cpp"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    cpp_result_efficient = subprocess.run(
        ["./max_pairwise_product_super_efficient"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    # Ejecutar el código Python y capturar salida
    py_result = subprocess.run(
        ["python3", "max_pairwise_product_sort.py"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()

    # Ejecutar el código Python y capturar salida
    py_result_efficient = subprocess.run(
        ["python3", "max_pairwise_product_super_efficient.py"],
        input=input_data,
        text=True,
        capture_output=True
    ).stdout.strip()


    # Validar que los resultados sean iguales
    if int(cpp_result) != int(cpp_result_efficient) or int(py_result) != int(py_result_efficient):
        print(f"❌ Test {test_id + 1} failed!")
        print(f"Test {test_id + 1}:")
        print(f"Input data: {input_data.strip()}")
        print(f"C++ Output: {cpp_result}")
        print(f"Python Output: {py_result}")
        print(f"Python Output efficient: {py_result_efficient}")
        print(f"C++ Output efficient: {cpp_result_efficient}\n")
        break

print("✅ All tests passed!")

✅ All tests passed!
